In [1]:
#pip install numpy pandas geopandas shapely osmnx requests

In [2]:
#pip install --upgrade osmnx

In [3]:
#pip install osmnx geopandas

In [4]:
#pip install geopy

# Berlin missions - Data transformation Overview

This notebook transforms raw diplomatic mission data from OpenStreetMap into a structured relational table for the final database.
It is part of the `berlin_source_data` schema and represents embassies, consulates, and honorary consulates.

The transformation includes:
- Cleaning missing or inconsistent values
- Standardizing district identifiers
- Validating ISO country codes
- Ensuring referential integrity with the districts table
- Preparing geospatial coordinates (latitude / longitude)

The result is the table:
`berlin_source_data.diplomatic_missions`

## Purpose

The table is used to store geospatial and administrative information about diplomatic missions, including their location, district, and contact details.

## Columns

| Column Name        | Data Type        | Description |
|--------------------|------------------|-------------|
| id                 | VARCHAR(20)      | Unique identifier of the diplomatic mission (Primary Key) |
| district_id        | VARCHAR(20)      | Identifier of the Berlin district (Foreign Key) |
| name               | VARCHAR(200)     | Official name of the diplomatic mission |
| latitude           | DECIMAL(9,6)     | Latitude coordinate of the mission location |
| longitude          | DECIMAL(9,6)     | Longitude coordinate of the mission location |
| geometry           | VARCHAR          | Geospatial representation (e.g., WKT format) |
| neighborhood       | VARCHAR(100)     | Name of the neighborhood |
| district           | VARCHAR(100)     | Name of the district |
| neighborhood_id    | VARCHAR(20)      | Identifier of the neighborhood |
| country_iso_code   | CHAR(2)          | ISO 3166-1 alpha-2 country code (e.g., 'US', 'FR') |
| mission_type       | VARCHAR(50)      | Type of mission (Embassy, Consulate, Honorary Consulate) |
| address            | VARCHAR(200)     | Street address |
| website            | VARCHAR(200)     | Official website |
| email              | VARCHAR(100)     | Contact email address |

## Constraints

- **Primary Key:** `id`
- **Foreign Key:** `district_id` references `berlin_source_data.districts(district_id)`
- `district_id` is mandatory (`NOT NULL`)
- Default value for `name` is `'Unknown'`
- Referential integrity is enforced:
  - `ON DELETE RESTRICT`
  - `ON UPDATE CASCADE`

## Data Relationships

The table is linked to the `districts` table via `district_id`.  
Each diplomatic mission belongs to exactly one district.



# Import libraries

In [5]:
import numpy as np
import pandas as pd
import geopandas as gpd
from shapely.geometry import Point, Polygon
import osmnx as ox
import requests
import json
from geopy.geocoders import Nominatim
from geopy.extra.rate_limiter import RateLimiter

import psycopg2
from sqlalchemy import create_engine, text
import warnings

warnings.filterwarnings("ignore")

# Load data

**Evaluated but Not Integrated**
- Wikidata  
- German Federal Foreign Office (Auswärtiges Amt)

Wikidata and the official Foreign Office list were reviewed and compared against OSM.  
Due to negligible overlap improvements and no critical missing attributes in the OSM dataset, a merge was not considered meaningful for the MVP.  
OSM therefore remains the sole integrated source.

In [6]:
place = "Berlin, Germany"

# Fetch diplomatic features from OSM via Overpass (wrapped by OSMnx)
gdf = ox.features_from_place(
    place,
    tags={"office": "diplomatic", "diplomatic": True}
)

# Keep only selected diplomatic types (equivalent to the Overpass regex)
allowed = {
    "embassy",
    "consulate",
    "consulate_general",
    "nunciature",
    "permanent_mission",
    "mission",
}
gdf = gdf[gdf["diplomatic"].isin(allowed)].copy()

# Create representative coordinates (centroid for non-point geometries)
rep = gdf.geometry.representative_point()
gdf["lat"] = rep.y
gdf["lon"] = rep.x

# Convert to a plain DataFrame (drop geometry if not needed downstream)
df_osm = gdf.reset_index()

df_osm

,element,id,geometry,addr:city,addr:housenumber,addr:postcode,addr:street,check_date:opening_hours,contact:email,contact:fax,...,contact:country,contact:housenumber,contact:postcode,contact:street,contact:suburb,note:addr,short_name:it,description:de,lat,lon
0,node,87040412,POINT (13.38432 52.51144),Berlin,44,10117,Wilhelmstraße,2025-06-11,berlin@embassy.mzv.cz,+49 30 2294033,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,52.511439,13.384316
1,node,89304553,POINT (13.40936 52.51227),Berlin,NaN,10179,NaN,2025-04-13,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,52.512265,13.409365
2,node,348075887,POINT (13.31448 52.47575),Berlin,16,14197,Hohensteiner Straße,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,52.475747,13.314478
3,node,361488016,POINT (13.41363 52.51362),Berlin,57,10179,Wallstraße,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,52.513618,13.413626
4,node,361488018,POINT (13.4168 52.51361),Berlin,54,10179,Märkisches Ufer,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,52.513610,13.416802
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
210,way,1162427498,"POLYGON ((13.2304 52.4339, 13.23055 52.43359, ...",Berlin,19,14163,Niklasstraße,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,52.433745,13.230721
211,way,1201706863,"POLYGON ((13.35638 52.50927, 13.35655 52.50961...",NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,52.509404,13.356684
212,way,1241734777,"POLYGON ((13.35904 52.50838, 13.35863 52.50847...",Berlin,11-15,10785,Hiroshimastraße,NaN,gremb.ber@mfa.gr,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,52.508568,13.358883
213,way,1251105122,"POLYGON ((13.35218 52.50158, 13.35182 52.50154...",NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,52.501356,13.352032


In [7]:
print(df_osm.columns)

Index(['element', 'id', 'geometry', 'addr:city', 'addr:housenumber',
       'addr:postcode', 'addr:street', 'check_date:opening_hours',
       'contact:email', 'contact:fax',
       ...
       'contact:country', 'contact:housenumber', 'contact:postcode',
       'contact:street', 'contact:suburb', 'note:addr', 'short_name:it',
       'description:de', 'lat', 'lon'],
      dtype='object', length=250)


## Columnmapping

| Target column | OSM column(s)	| Transform |
|---------|--------|---------| 
| id	| Varchar(20) | Primary Key, numeric only, no letters |
| source_id	| type, id	| f"{type}/{id}" | 
| data_source	| -	| "osm" | 
| name	| name	| fallback 'Unknown' | 
| latitude	| lat	| to numeric | 
| longitude	| lon	| to numeric | 
| geometry	| lat, lon	| POINT(lon lat) | 
| mission_type	| office/diplomatic/consulate	| map tags -> category | 
| country	| country	| keep as is (optional normalize later) | 
| address	| addr:street, addr:housenumber, addr:postcode, addr:city	| concat to single string | 
| website*	| website/contact:website	| choose first non-null | 
| email*	| email/contact:email	| choose first non-null | 

## Transform OSM data

In [8]:
print(df_osm.shape)

(215, 250)


In [9]:
# Build GeoDataFrame from lon/lat
gdf_osm = gpd.GeoDataFrame(
    df_osm.copy(),
    geometry=gpd.points_from_xy(df_osm["lon"], df_osm["lat"]),
    crs="EPSG:4326"
)

gdf_osm

,element,id,geometry,addr:city,addr:housenumber,addr:postcode,addr:street,check_date:opening_hours,contact:email,contact:fax,...,contact:country,contact:housenumber,contact:postcode,contact:street,contact:suburb,note:addr,short_name:it,description:de,lat,lon
0,node,87040412,POINT (13.38432 52.51144),Berlin,44,10117,Wilhelmstraße,2025-06-11,berlin@embassy.mzv.cz,+49 30 2294033,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,52.511439,13.384316
1,node,89304553,POINT (13.40936 52.51227),Berlin,NaN,10179,NaN,2025-04-13,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,52.512265,13.409365
2,node,348075887,POINT (13.31448 52.47575),Berlin,16,14197,Hohensteiner Straße,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,52.475747,13.314478
3,node,361488016,POINT (13.41363 52.51362),Berlin,57,10179,Wallstraße,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,52.513618,13.413626
4,node,361488018,POINT (13.4168 52.51361),Berlin,54,10179,Märkisches Ufer,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,52.513610,13.416802
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
210,way,1162427498,POINT (13.23072 52.43375),Berlin,19,14163,Niklasstraße,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,52.433745,13.230721
211,way,1201706863,POINT (13.35668 52.5094),NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,52.509404,13.356684
212,way,1241734777,POINT (13.35888 52.50857),Berlin,11-15,10785,Hiroshimastraße,NaN,gremb.ber@mfa.gr,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,52.508568,13.358883
213,way,1251105122,POINT (13.35203 52.50136),NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,52.501356,13.352032


In [10]:
# Keep only the columns we need from OSM
osm_cols = [
    "id","type","lat","lon", "geometry", "name","office","diplomatic","consulate","country",
    "addr:street","addr:housenumber","addr:postcode","addr:city",
    "website","contact:website",
    "phone","contact:phone","email","contact:email"
]

missions_raw = gdf_osm[[c for c in osm_cols if c in gdf_osm.columns]].copy()
missions_raw.head()

,id,type,lat,lon,geometry,name,office,diplomatic,consulate,country,addr:street,addr:housenumber,addr:postcode,addr:city,website,contact:website,phone,contact:phone,email,contact:email
0,87040412,NaN,52.511439,13.384316,POINT (13.38432 52.51144),Botschaft der Tschechischen Republik,diplomatic,embassy,NaN,CZ,Wilhelmstraße,44,10117,Berlin,https://www.mzv.cz/berlin,NaN,NaN,+49 30 226380,NaN,berlin@embassy.mzv.cz
1,89304553,NaN,52.512265,13.409365,POINT (13.40936 52.51227),Botschaft von Australien,diplomatic,embassy,NaN,AU,NaN,NaN,10179,Berlin,https://germany.embassy.gov.au/,NaN,+49 30 88 00 88 0,NaN,info.berlin@dfat.gov.au,NaN
2,348075887,NaN,52.475747,13.314478,POINT (13.31448 52.47575),Botschaft Gabun,diplomatic,embassy,NaN,GA,Hohensteiner Straße,16,14197,Berlin,http://www.amba-allemagne.ga/,NaN,NaN,NaN,NaN,NaN
3,361488016,NaN,52.513618,13.413626,POINT (13.41363 52.51362),Brasilianische Botschaft,diplomatic,embassy,NaN,BR,Wallstraße,57,10179,Berlin,http://berlim.itamaraty.gov.br/de/,NaN,+49 30 72628,NaN,NaN,NaN
4,361488018,NaN,52.513610,13.416802,POINT (13.4168 52.51361),Botschaft der Volksrepublik China,diplomatic,embassy,NaN,CN,Märkisches Ufer,54,10179,Berlin,NaN,http://www.china-botschaft.de/det/,NaN,NaN,NaN,NaN


In [11]:
# Helper function to normalize string values:
# - converts NaN to None
# - trims whitespace
# - replaces empty strings with None
def clean_str(x):
    if pd.isna(x):
        return None
    s = str(x).strip()
    return s if s else None

missions = missions_raw.copy()

# Standard lat/lon names for the DB schema
missions["latitude"] = pd.to_numeric(missions["lat"], errors="coerce")
missions["longitude"] = pd.to_numeric(missions["lon"], errors="coerce")

# Prefer contact:website over website
missions["website"] = missions.get("contact:website").fillna(missions.get("website"))
missions["website"] = missions["website"].apply(clean_str).astype("string")

# Prefer contact:email over email
missions["email"] = missions.get("contact:email").fillna(missions.get("email"))
missions["email"] = missions["email"].apply(clean_str).astype("string")

# Full address string
def build_address(row):
    street = clean_str(row.get("addr:street"))
    hn = clean_str(row.get("addr:housenumber"))
    pc = clean_str(row.get("addr:postcode"))  
    city = clean_str(row.get("addr:city")) or "Berlin"

    line1 = " ".join([p for p in [street, hn] if p])
    if line1 and pc:
        return f"{line1}, {pc} {city}"
    if line1:
        return f"{line1}, {city}"
    if pc:
        return f"{pc} {city}"
    return None

missions["address"] = missions.apply(build_address, axis=1)


# Mission type mapping (MVP normalization)
def map_mission_type(x):
    x = clean_str(x)
    if not x:
        return None
    x = x.lower()
    if x == "embassy":
        return "embassy"
    if x == "nunciature":
        return "nunciature"
    if x in ["consulate", "consulate_general", "permanent_mission", "mission"]:
        return "consulate"
    if x in ["honorary_consulate", "consulate_honorary"]:
        return "honorary_consulate"
    return None

missions["mission_type"] = missions.get("diplomatic").apply(map_mission_type)

# Country (OSM column is already available)
missions["country_iso_code"] = missions.get("country").apply(clean_str)

# Source label
missions["data_source"] = "OSM"

# Create a stable order so IDs don't change between runs
missions = missions.sort_values(
    by=["name", "diplomatic", "lat", "lon"],
    na_position="last"
).reset_index(drop=True)

missions[["id","name","mission_type","country_iso_code","address","email","website"]].head(20)

,id,name,mission_type,country_iso_code,address,email,website
0,78061104,Afghanisches Konsulat,consulate,AF,None,<NA>,<NA>
1,39348994,Apostolische Nuntiatur in Deutschland,embassy,VA,"Lilienthalstraße 3a, 10965 Berlin",<NA>,http://www.nuntiatur.de
2,1880606692,Botschaft Argentinien,embassy,AR,"Kleiststraße 23-26, 10787 Berlin",<NA>,http://www.ealem.mrecic.gov.ar/
3,348075887,Botschaft Gabun,embassy,GA,"Hohensteiner Straße 16, 14197 Berlin",<NA>,http://www.amba-allemagne.ga/
4,40664508,Botschaft Jordanien,embassy,JO,"Heerstraße 201, 13595 Berlin",<NA>,http://www.jordanembassy.de
5,76775131,Botschaft Lettlands,embassy,LV,"Reinerzstraße 40-41, 14193 Berlin",embassy.germany@mfa.gov.lv,https://www.mfa.gov.lv/de
6,123811011,Botschaft Tadschikistan,embassy,TJ,"Perleberger Straße 43, 10559 Berlin",info@botschaft-tadschikistan.de,http://www.botschaft-tadschikistan.de
7,25999440,Botschaft der Arabischen Republik Syrien,embassy,SY,"Rauchstraße 25, 10787 Berlin",<NA>,http://www.mofaex.gov.sy/berlin-embassy
8,24045732,Botschaft der Arabischen Republik Ägypten,embassy,EG,"Stauffenbergstraße 6-7, 10785 Berlin",embassy@egyptian-embassy.de,https://egyptian-embassy.de/
9,8615201458,Botschaft der Bolivarischen Republik Venezuela,embassy,VE,"Schillstraße 10, 10785 Berlin",<NA>,https://www.botschaft-venezuela.de


In [12]:
# Load official Berlin districts GeoDataFrame from lor_ortsteile.geojson
lor_gdf = gpd.read_file("lor_ortsteile.geojson")

# Ensure same CRS
if lor_gdf.crs != missions.crs:
    lor_gdf = lor_gdf.to_crs(missions.crs)

In [13]:
# Spatial join: matching your cinemas with district and neighborhoods
missions_enriched = gpd.sjoin(
    missions,
    lor_gdf[["BEZIRK", "spatial_name", "OTEIL","geometry"]],
    how="left",
    predicate="within"
)
missions_enriched.head()

,id,type,lat,lon,geometry,name,office,diplomatic,consulate,country,...,latitude,longitude,address,mission_type,country_iso_code,data_source,index_right,BEZIRK,spatial_name,OTEIL
0,78061104,NaN,52.479652,13.278615,POINT (13.27862 52.47965),Afghanisches Konsulat,diplomatic,consulate,yes,AF,...,52.479652,13.278615,None,consulate,AF,OSM,24,Charlottenburg-Wilmersdorf,0404,Grunewald
1,39348994,NaN,52.488004,13.408922,POINT (13.40892 52.488),Apostolische Nuntiatur in Deutschland,diplomatic,embassy,NaN,VA,...,52.488004,13.408922,"Lilienthalstraße 3a, 10965 Berlin",embassy,VA,OSM,50,Neukölln,0801,Neukölln
2,1880606692,NaN,52.501211,13.344571,POINT (13.34457 52.50121),Botschaft Argentinien,diplomatic,embassy,NaN,AR,...,52.501211,13.344571,"Kleiststraße 23-26, 10787 Berlin",embassy,AR,OSM,44,Tempelhof-Schöneberg,0701,Schöneberg
3,348075887,NaN,52.475747,13.314478,POINT (13.31448 52.47575),Botschaft Gabun,diplomatic,embassy,NaN,GA,...,52.475747,13.314478,"Hohensteiner Straße 16, 14197 Berlin",embassy,GA,OSM,22,Charlottenburg-Wilmersdorf,0402,Wilmersdorf
4,40664508,NaN,52.511867,13.201596,POINT (13.2016 52.51187),Botschaft Jordanien,diplomatic,embassy,NaN,JO,...,52.511867,13.201596,"Heerstraße 201, 13595 Berlin",embassy,JO,OSM,36,Spandau,0509,Wilhelmstadt


In [14]:
# Rename boundary fields to match the POI schema
missions_enriched = missions_enriched.rename(columns={
    "BEZIRK": "district",
    "OTEIL": "neighborhood",
    "spatial_name": "neighborhood_id"
}).drop(columns=["index_right"], errors="ignore")

missions_enriched[["name", "district", "neighborhood", "neighborhood_id"]].head(10)

,name,district,neighborhood,neighborhood_id
0,Afghanisches Konsulat,Charlottenburg-Wilmersdorf,Grunewald,0404
1,Apostolische Nuntiatur in Deutschland,Neukölln,Neukölln,0801
2,Botschaft Argentinien,Tempelhof-Schöneberg,Schöneberg,0701
3,Botschaft Gabun,Charlottenburg-Wilmersdorf,Wilmersdorf,0402
4,Botschaft Jordanien,Spandau,Wilhelmstadt,0509
5,Botschaft Lettlands,Charlottenburg-Wilmersdorf,Schmargendorf,0403
6,Botschaft Tadschikistan,Mitte,Moabit,0102
7,Botschaft der Arabischen Republik Syrien,Mitte,Tiergarten,0104
8,Botschaft der Arabischen Republik Ägypten,Mitte,Tiergarten,0104
9,Botschaft der Bolivarischen Republik Venezuela,Mitte,Tiergarten,0104


In [15]:
# District id mapping
district_mapping = {
    'Mitte': '11001001',
    'Friedrichshain-Kreuzberg': '11002002',
    'Pankow': '11003003',
    'Charlottenburg-Wilmersdorf': '11004004',
    'Spandau': '11005005',
    'Steglitz-Zehlendorf': '11006006',
    'Tempelhof-Schöneberg': '11007007',
    'Neukölln': '11008008',
    'Treptow-Köpenick': '11009009',
    'Marzahn-Hellersdorf': '11010010',
    'Lichtenberg': '11011011',
    'Reinickendorf': '11012012'
}

# Apply mapping to create district_id column (string)
missions_enriched["district_id"] = missions_enriched["district"].map(district_mapping).astype(str)

missions_enriched[["name","district","district_id","neighborhood","neighborhood_id"]].head(5)
#missions_enriched[["id","name","mission_type","country_iso_code","address","email","website"]].head(20)

,name,district,district_id,neighborhood,neighborhood_id
0,Afghanisches Konsulat,Charlottenburg-Wilmersdorf,11004004,Grunewald,0404
1,Apostolische Nuntiatur in Deutschland,Neukölln,11008008,Neukölln,0801
2,Botschaft Argentinien,Tempelhof-Schöneberg,11007007,Schöneberg,0701
3,Botschaft Gabun,Charlottenburg-Wilmersdorf,11004004,Wilmersdorf,0402
4,Botschaft Jordanien,Spandau,11005005,Wilhelmstadt,0509


In [16]:
missions_enriched["geometry"] = missions_enriched.apply(
    lambda r: f"POINT({r['longitude']} {r['latitude']})",
    axis=1
)

In [17]:
TARGET_COLUMNS = [
    "id",
    "district_id",
    "name",
    "latitude",
    "longitude",
    "geometry",
    "neighborhood",
    "district",
    "neighborhood_id",
    "country_iso_code",
    "mission_type",
    "address",
    "website",
    "email",
]

df_missions_final = missions_enriched[TARGET_COLUMNS].copy()
df_missions_final.head()

,id,district_id,name,latitude,longitude,geometry,neighborhood,district,neighborhood_id,country_iso_code,mission_type,address,website,email
0,78061104,11004004,Afghanisches Konsulat,52.479652,13.278615,POINT(13.27861531770446 52.47965235),Grunewald,Charlottenburg-Wilmersdorf,0404,AF,consulate,None,<NA>,<NA>
1,39348994,11008008,Apostolische Nuntiatur in Deutschland,52.488004,13.408922,POINT(13.408922272633824 52.488003649999996),Neukölln,Neukölln,0801,VA,embassy,"Lilienthalstraße 3a, 10965 Berlin",http://www.nuntiatur.de,<NA>
2,1880606692,11007007,Botschaft Argentinien,52.501211,13.344571,POINT(13.3445706 52.5012109),Schöneberg,Tempelhof-Schöneberg,0701,AR,embassy,"Kleiststraße 23-26, 10787 Berlin",http://www.ealem.mrecic.gov.ar/,<NA>
3,348075887,11004004,Botschaft Gabun,52.475747,13.314478,POINT(13.314478 52.4757469),Wilmersdorf,Charlottenburg-Wilmersdorf,0402,GA,embassy,"Hohensteiner Straße 16, 14197 Berlin",http://www.amba-allemagne.ga/,<NA>
4,40664508,11005005,Botschaft Jordanien,52.511867,13.201596,POINT(13.201596053308027 52.5118667),Wilhelmstadt,Spandau,0509,JO,embassy,"Heerstraße 201, 13595 Berlin",http://www.jordanembassy.de,<NA>


In [18]:
df_missions_final.info()

<class 'pandas.core.frame.DataFrame'>
Index: 215 entries, 0 to 214
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   id                215 non-null    int64  
 1   district_id       215 non-null    object 
 2   name              208 non-null    object 
 3   latitude          215 non-null    float64
 4   longitude         215 non-null    float64
 5   geometry          215 non-null    object 
 6   neighborhood      215 non-null    object 
 7   district          215 non-null    object 
 8   neighborhood_id   215 non-null    object 
 9   country_iso_code  208 non-null    object 
 10  mission_type      215 non-null    object 
 11  address           154 non-null    object 
 12  website           163 non-null    string 
 13  email             69 non-null     string 
dtypes: float64(2), int64(1), object(9), string(2)
memory usage: 25.2+ KB


# drop duplicates

In [19]:
df_missions_final= df_missions_final.drop_duplicates(subset=["id"])

In [20]:
df_missions_final["name"].value_counts().head(20)

name
Honorargeneralkonsul des Königreichs Eswatini    3
Botschaft der Mongolei                           2
Botschaft der Republik Jemen                     2
Botschaft von Sri Lanka                          2
Botschaft von Tansania                           1
Botschaft von Laos                               1
Botschaft von Malaysia                           1
Botschaft von Montenegro                         1
Botschaft von Nepal                              1
Botschaft von Neuseeland                         1
Botschaft von Portugal                           1
Botschaft von Rumänien                           1
Botschaft von Somalia                            1
Botschaft von Turkmenistan                       1
Botschaft von Japan                              1
Botschaft von Ungarn                             1
Botschaft Äthiopien                              1
Botschaftsresidenz von Irak                      1
Botschaftsresidenz von Äthiopien                 1
Brasilianische Botschaft  

In [21]:
# Identify names that appear more than once
duplicate_names = (
    df_missions_final["name"]
    .value_counts()
)

duplicate_names = duplicate_names[duplicate_names > 1].index

# Show all rows with those duplicate names
df_missions_final[
    df_missions_final["name"].isin(duplicate_names)
].sort_values("name")

,id,district_id,name,latitude,longitude,geometry,neighborhood,district,neighborhood_id,country_iso_code,mission_type,address,website,email
22,3885026854,11001001,Botschaft der Mongolei,52.513334,13.397533,POINT(13.3975334 52.5133337),Mitte,Mitte,0101,MN,embassy,"Hausvogteiplatz 14, 10117 Berlin",http://www.berlin.embassy.mn/,<NA>
23,8907796654,11003003,Botschaft der Mongolei,52.582708,13.402478,POINT(13.4024781 52.5827084),Niederschönhausen,Pankow,0311,MN,embassy,"Dietzgenstraße 31, Berlin",<NA>,<NA>
44,638521475,11006006,Botschaft der Republik Jemen,52.458685,13.311531,POINT(13.3115312 52.4586847),Steglitz,Steglitz-Zehlendorf,0601,YE,embassy,None,http://yemenembassy.de/,<NA>
45,700080648,11004004,Botschaft der Republik Jemen,52.504902,13.340898,POINT(13.3408976 52.5049018),Charlottenburg,Charlottenburg-Wilmersdorf,0401,YE,embassy,"Budapester Straße 37, 10787 Berlin",http://yemenembassy.de,info@botschaft-jemen.de
141,1162427498,11006006,Botschaft von Sri Lanka,52.433745,13.230721,POINT(13.230720560437014 52.4337453),Zehlendorf,Steglitz-Zehlendorf,0604,LK,embassy,"Niklasstraße 19, 14163 Berlin",https://www.srilanka-botschaft.de,<NA>
142,118935982,11006006,Botschaft von Sri Lanka,52.433776,13.230691,POINT(13.230690656821647 52.43377555),Zehlendorf,Steglitz-Zehlendorf,0604,None,embassy,"Niklasstraße 19, 14163 Berlin",https://srilanka-botschaft.de/,<NA>
155,10738288360,11004004,Honorargeneralkonsul des Königreichs Eswatini,52.495651,13.308071,POINT(13.308071 52.4956509),Wilmersdorf,Charlottenburg-Wilmersdorf,0402,None,consulate,None,<NA>,<NA>
156,6373769056,11001001,Honorargeneralkonsul des Königreichs Eswatini,52.514802,13.387434,POINT(13.3874336 52.5148024),Mitte,Mitte,0101,SZ,consulate,None,<NA>,<NA>
157,1820157445,11001001,Honorargeneralkonsul des Königreichs Eswatini,52.522954,13.400570,POINT(13.40057 52.5229536),Mitte,Mitte,0101,SZ,consulate,None,https://www.swasiland.de/konsulat/generalkonsu...,<NA>


In [22]:
# extract rows
sl = df_missions_final[
    df_missions_final["name"] == "Botschaft von Sri Lanka"
].copy()

# build new combined row
row1 = sl.iloc[0]
row2 = sl.iloc[1]

merged_row = row1.copy()

# Fill missing values from second row
for col in df_missions_final.columns:
    if pd.isna(merged_row[col]) and not pd.isna(row2[col]):
        merged_row[col] = row2[col]

merged_row["latitude"] = sl["latitude"].mean()
merged_row["longitude"] = sl["longitude"].mean()

# Remove old rows
df_missions_final = df_missions_final[
    df_missions_final["name"] != "Botschaft von Sri Lanka"
]

# Append merged row
df_missions_final = pd.concat(
    [df_missions_final, pd.DataFrame([merged_row])],
    ignore_index=True
)

In [23]:
df_missions_final[
    df_missions_final["name"] == "Botschaft von Sri Lanka"
]

,id,district_id,name,latitude,longitude,geometry,neighborhood,district,neighborhood_id,country_iso_code,mission_type,address,website,email
213,1162427498,11006006,Botschaft von Sri Lanka,52.43376,13.230706,POINT(13.230720560437014 52.4337453),Zehlendorf,Steglitz-Zehlendorf,0604,LK,embassy,"Niklasstraße 19, 14163 Berlin",https://www.srilanka-botschaft.de,<NA>


**Duplicate Assessment Summary**
- Mongolei: No duplicate. The entries represent distinct locations and were retained.
- Jemen: No duplicate. The entries refer to different locations and were retained.
- Sri Lanka: Identified as a technical duplicate (same address and nearly identical coordinates). The records were merged to one row.
- Eswatini: No coordinate-based duplicate detected. Although multiple entries exist, no clear technical duplication could be verified; therefore, all records were retained.

In [24]:
df_missions_final["name"].value_counts().head(20)

name
Honorargeneralkonsul des Königreichs Eswatini    3
Botschaft der Mongolei                           2
Botschaft der Republik Jemen                     2
Afghanisches Konsulat                            1
Botschaft von Turkmenistan                       1
Botschaft von Laos                               1
Botschaft von Malaysia                           1
Botschaft von Montenegro                         1
Botschaft von Nepal                              1
Botschaft von Neuseeland                         1
Botschaft von Portugal                           1
Botschaft von Rumänien                           1
Botschaft von Somalia                            1
Botschaft von Tansania                           1
Botschaft von Ungarn                             1
Botschaft von Japan                              1
Botschaft Äthiopien                              1
Botschaftsresidenz von Irak                      1
Botschaftsresidenz von Äthiopien                 1
Brasilianische Botschaft  

# Missing values

In [25]:
df_missions_final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 214 entries, 0 to 213
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   id                214 non-null    int64  
 1   district_id       214 non-null    object 
 2   name              207 non-null    object 
 3   latitude          214 non-null    float64
 4   longitude         214 non-null    float64
 5   geometry          214 non-null    object 
 6   neighborhood      214 non-null    object 
 7   district          214 non-null    object 
 8   neighborhood_id   214 non-null    object 
 9   country_iso_code  208 non-null    object 
 10  mission_type      214 non-null    object 
 11  address           153 non-null    object 
 12  website           162 non-null    object 
 13  email             69 non-null     string 
dtypes: float64(2), int64(1), object(10), string(1)
memory usage: 23.5+ KB


## missing names

In [26]:
# missing names
df_missions_final[df_missions_final["name"].isna()]

,id,district_id,name,latitude,longitude,geometry,neighborhood,district,neighborhood_id,country_iso_code,mission_type,address,website,email
206,11091458818,11006006,NaN,52.470303,13.289067,POINT(13.2890666 52.4703035),Dahlem,Steglitz-Zehlendorf,0605,Kenia,embassy,"Rheinbabenallee 49, 14199 Berlin",https://kenyaembassyberlin.de,<NA>
207,11687769898,11004004,NaN,52.477151,13.274898,POINT(13.2748984 52.477151),Grunewald,Charlottenburg-Wilmersdorf,0404,Luxembourg,embassy,"Wildpfad 5, 14193 Berlin",<NA>,<NA>
208,12541996229,11004004,NaN,52.492526,13.285668,POINT(13.2856685 52.4925256),Grunewald,Charlottenburg-Wilmersdorf,0404,Haiti,embassy,"Kunz-Buntschuh-Straße 13, 14193 Berlin",<NA>,<NA>
209,5580806145,11001001,NaN,52.504640,13.343969,POINT(13.3439691 52.5046401),Tiergarten,Mitte,0104,DJ,embassy,None,http://www.dschibuti-botschaft.de/,<NA>
210,5580806144,11001001,NaN,52.504692,13.343829,POINT(13.343829 52.5046918),Tiergarten,Mitte,0104,LS,embassy,None,https://lesothoembassy.de/,info@lesothoembassy.de
211,64384567,11003003,NaN,52.555918,13.411762,POINT(13.411761862347964 52.55591785),Prenzlauer Berg,Pankow,0301,ER,embassy,None,<NA>,<NA>
212,76978937,11012012,NaN,52.605031,13.221587,POINT(13.221587211175095 52.6050307),Heiligensee,Reinickendorf,1204,Kongo,embassy,13503 Berlin,<NA>,<NA>


In [27]:
mask = (
    df_missions_final["name"].isna() &
    df_missions_final["mission_type"].eq("embassy")
)

df_missions_final.loc[mask, "name"] = (
    "Botschaft von " + df_missions_final.loc[mask, "country_iso_code"]
)

In [28]:
df_missions_final[df_missions_final["name"].isna()]

,id,district_id,name,latitude,longitude,geometry,neighborhood,district,neighborhood_id,country_iso_code,mission_type,address,website,email


In [29]:
df_missions_final

,id,district_id,name,latitude,longitude,geometry,neighborhood,district,neighborhood_id,country_iso_code,mission_type,address,website,email
0,78061104,11004004,Afghanisches Konsulat,52.479652,13.278615,POINT(13.27861531770446 52.47965235),Grunewald,Charlottenburg-Wilmersdorf,0404,AF,consulate,None,<NA>,<NA>
1,39348994,11008008,Apostolische Nuntiatur in Deutschland,52.488004,13.408922,POINT(13.408922272633824 52.488003649999996),Neukölln,Neukölln,0801,VA,embassy,"Lilienthalstraße 3a, 10965 Berlin",http://www.nuntiatur.de,<NA>
2,1880606692,11007007,Botschaft Argentinien,52.501211,13.344571,POINT(13.3445706 52.5012109),Schöneberg,Tempelhof-Schöneberg,0701,AR,embassy,"Kleiststraße 23-26, 10787 Berlin",http://www.ealem.mrecic.gov.ar/,<NA>
3,348075887,11004004,Botschaft Gabun,52.475747,13.314478,POINT(13.314478 52.4757469),Wilmersdorf,Charlottenburg-Wilmersdorf,0402,GA,embassy,"Hohensteiner Straße 16, 14197 Berlin",http://www.amba-allemagne.ga/,<NA>
4,40664508,11005005,Botschaft Jordanien,52.511867,13.201596,POINT(13.201596053308027 52.5118667),Wilhelmstadt,Spandau,0509,JO,embassy,"Heerstraße 201, 13595 Berlin",http://www.jordanembassy.de,<NA>
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
209,5580806145,11001001,Botschaft von DJ,52.504640,13.343969,POINT(13.3439691 52.5046401),Tiergarten,Mitte,0104,DJ,embassy,None,http://www.dschibuti-botschaft.de/,<NA>
210,5580806144,11001001,Botschaft von LS,52.504692,13.343829,POINT(13.343829 52.5046918),Tiergarten,Mitte,0104,LS,embassy,None,https://lesothoembassy.de/,info@lesothoembassy.de
211,64384567,11003003,Botschaft von ER,52.555918,13.411762,POINT(13.411761862347964 52.55591785),Prenzlauer Berg,Pankow,0301,ER,embassy,None,<NA>,<NA>
212,76978937,11012012,Botschaft von Kongo,52.605031,13.221587,POINT(13.221587211175095 52.6050307),Heiligensee,Reinickendorf,1204,Kongo,embassy,13503 Berlin,<NA>,<NA>


In [30]:
df_missions_final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 214 entries, 0 to 213
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   id                214 non-null    int64  
 1   district_id       214 non-null    object 
 2   name              214 non-null    object 
 3   latitude          214 non-null    float64
 4   longitude         214 non-null    float64
 5   geometry          214 non-null    object 
 6   neighborhood      214 non-null    object 
 7   district          214 non-null    object 
 8   neighborhood_id   214 non-null    object 
 9   country_iso_code  208 non-null    object 
 10  mission_type      214 non-null    object 
 11  address           153 non-null    object 
 12  website           162 non-null    object 
 13  email             69 non-null     string 
dtypes: float64(2), int64(1), object(10), string(1)
memory usage: 23.5+ KB


## missing country

In [31]:
# missing country
df_missions_final[df_missions_final["country_iso_code"].isna()]

,id,district_id,name,latitude,longitude,geometry,neighborhood,district,neighborhood_id,country_iso_code,mission_type,address,website,email
13,12883500820,11004004,Botschaft der Dominikanischen Republik,52.501061,13.322651,POINT(13.322651 52.5010614),Charlottenburg,Charlottenburg-Wilmersdorf,0401,None,embassy,"Knesebeckstraße 61, 10719 Berlin",https://deu.mirex.gob.do/,infoembalemania@mirex.gob.do
128,1201706863,11001001,Botschaft von Indonesien,52.509404,13.356684,POINT(13.356683580864473 52.50940395),Tiergarten,Mitte,0104,None,embassy,None,<NA>,<NA>
149,13432587157,11004004,Embassy of the Republic of Mauritius,52.499692,13.292917,POINT(13.2929167 52.4996921),Halensee,Charlottenburg-Wilmersdorf,0407,None,embassy,None,<NA>,<NA>
153,10738288360,11004004,Honorargeneralkonsul des Königreichs Eswatini,52.495651,13.308071,POINT(13.308071 52.4956509),Wilmersdorf,Charlottenburg-Wilmersdorf,0402,None,consulate,None,<NA>,<NA>
158,12158159241,11006006,Iran Haus - Kulturabteilung der Botschaft der ...,52.432508,13.308956,POINT(13.308956 52.4325077),Lichterfelde,Steglitz-Zehlendorf,0602,None,embassy,None,<NA>,<NA>
191,108631487,11006006,Residenz des Botschafters von Marokko,52.464230,13.282878,POINT(13.282877561101083 52.4642299),Dahlem,Steglitz-Zehlendorf,0605,None,embassy,"Im Dol 48, 14195 Berlin",<NA>,<NA>


In [32]:
df_missions_final[df_missions_final["country_iso_code"].isna()][
    ["id", "name"]
]

,id,name
13,12883500820,Botschaft der Dominikanischen Republik
128,1201706863,Botschaft von Indonesien
149,13432587157,Embassy of the Republic of Mauritius
153,10738288360,Honorargeneralkonsul des Königreichs Eswatini
158,12158159241,Iran Haus - Kulturabteilung der Botschaft der ...
191,108631487,Residenz des Botschafters von Marokko


In [33]:
for _id in [12883500820, 108631487,
            12158159241, 10738288360, 13432587157]:
    print(_id, (df_missions_final["id"] == _id).sum())

12883500820 1
108631487 1
12158159241 1
10738288360 1
13432587157 1


In [34]:
manual_iso_map = {
    12883500820: "DO",  # Dominikanische Republik
    108631487: "MA",  # Marokko
    12158159241: "IR",  # Iran
    1201706863: "ID",  # Indonesien
    13432587157: "MU",  # Mauritius
    10738288360: "SZ"	# Eswatini
}

for _id, country in manual_iso_map.items():
    df_missions_final.loc[
        df_missions_final["id"] == _id,
        "country_iso_code"
    ] = country

In [35]:
df_missions_final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 214 entries, 0 to 213
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   id                214 non-null    int64  
 1   district_id       214 non-null    object 
 2   name              214 non-null    object 
 3   latitude          214 non-null    float64
 4   longitude         214 non-null    float64
 5   geometry          214 non-null    object 
 6   neighborhood      214 non-null    object 
 7   district          214 non-null    object 
 8   neighborhood_id   214 non-null    object 
 9   country_iso_code  214 non-null    object 
 10  mission_type      214 non-null    object 
 11  address           153 non-null    object 
 12  website           162 non-null    object 
 13  email             69 non-null     string 
dtypes: float64(2), int64(1), object(10), string(1)
memory usage: 23.5+ KB


In [36]:
df_missions_final.head()

,id,district_id,name,latitude,longitude,geometry,neighborhood,district,neighborhood_id,country_iso_code,mission_type,address,website,email
0,78061104,11004004,Afghanisches Konsulat,52.479652,13.278615,POINT(13.27861531770446 52.47965235),Grunewald,Charlottenburg-Wilmersdorf,0404,AF,consulate,None,<NA>,<NA>
1,39348994,11008008,Apostolische Nuntiatur in Deutschland,52.488004,13.408922,POINT(13.408922272633824 52.488003649999996),Neukölln,Neukölln,0801,VA,embassy,"Lilienthalstraße 3a, 10965 Berlin",http://www.nuntiatur.de,<NA>
2,1880606692,11007007,Botschaft Argentinien,52.501211,13.344571,POINT(13.3445706 52.5012109),Schöneberg,Tempelhof-Schöneberg,0701,AR,embassy,"Kleiststraße 23-26, 10787 Berlin",http://www.ealem.mrecic.gov.ar/,<NA>
3,348075887,11004004,Botschaft Gabun,52.475747,13.314478,POINT(13.314478 52.4757469),Wilmersdorf,Charlottenburg-Wilmersdorf,0402,GA,embassy,"Hohensteiner Straße 16, 14197 Berlin",http://www.amba-allemagne.ga/,<NA>
4,40664508,11005005,Botschaft Jordanien,52.511867,13.201596,POINT(13.201596053308027 52.5118667),Wilhelmstadt,Spandau,0509,JO,embassy,"Heerstraße 201, 13595 Berlin",http://www.jordanembassy.de,<NA>


## missing addresses

In [37]:
# Geolocator
geolocator = Nominatim(
    user_agent="missions_address_restoration/1.0 (contact: max.mustermann@uni-imnetz.de)"
)

# RateLimiter (for OSM Guidelines)
reverse = RateLimiter(
    geolocator.reverse,
    min_delay_seconds=1.1,
    max_retries=2,
    error_wait_seconds=5.0
)

# Reverse-Function (structured + consistent format)
def safe_reverse(lat, lon):
    if pd.isna(lat) or pd.isna(lon):
        return None
    try:
        location = reverse((lat, lon), timeout=10, language="de")
        if not location:
            return None
        
        addr = location.raw.get("address", {})
        
        street = addr.get("road")
        house_number = addr.get("house_number")
        postcode = addr.get("postcode")
        city = addr.get("city") or addr.get("town") or "Berlin"
        
        line1 = " ".join([p for p in [street, house_number] if p])
        
        if line1 and postcode:
            return f"{line1}, {postcode} {city}"
        if line1:
            return f"{line1}, {city}"
        if postcode:
            return f"{postcode} {city}"
        return None
        
    except Exception as e:
        print("Reverse error:", type(e).__name__, e)
        return None

# don't overwrite existing values
mask = df_missions_final["address"].isna()

df_missions_final.loc[mask, "address"] = df_missions_final.loc[mask].apply(
    lambda row: safe_reverse(row["latitude"], row["longitude"]),
    axis=1
)


In [38]:
# control
print("Missing address values:", df_missions_final["address"].isna().sum())

# test
df_missions_final.head(20)

Missing address values: 0


,id,district_id,name,latitude,longitude,geometry,neighborhood,district,neighborhood_id,country_iso_code,mission_type,address,website,email
0,78061104,11004004,Afghanisches Konsulat,52.479652,13.278615,POINT(13.27861531770446 52.47965235),Grunewald,Charlottenburg-Wilmersdorf,0404,AF,consulate,"Kronberger Straße 5, 14193 Berlin",<NA>,<NA>
1,39348994,11008008,Apostolische Nuntiatur in Deutschland,52.488004,13.408922,POINT(13.408922272633824 52.488003649999996),Neukölln,Neukölln,0801,VA,embassy,"Lilienthalstraße 3a, 10965 Berlin",http://www.nuntiatur.de,<NA>
2,1880606692,11007007,Botschaft Argentinien,52.501211,13.344571,POINT(13.3445706 52.5012109),Schöneberg,Tempelhof-Schöneberg,0701,AR,embassy,"Kleiststraße 23-26, 10787 Berlin",http://www.ealem.mrecic.gov.ar/,<NA>
3,348075887,11004004,Botschaft Gabun,52.475747,13.314478,POINT(13.314478 52.4757469),Wilmersdorf,Charlottenburg-Wilmersdorf,0402,GA,embassy,"Hohensteiner Straße 16, 14197 Berlin",http://www.amba-allemagne.ga/,<NA>
4,40664508,11005005,Botschaft Jordanien,52.511867,13.201596,POINT(13.201596053308027 52.5118667),Wilhelmstadt,Spandau,0509,JO,embassy,"Heerstraße 201, 13595 Berlin",http://www.jordanembassy.de,<NA>
5,76775131,11004004,Botschaft Lettlands,52.484123,13.286223,POINT(13.28622318960423 52.48412305),Schmargendorf,Charlottenburg-Wilmersdorf,0403,LV,embassy,"Reinerzstraße 40-41, 14193 Berlin",https://www.mfa.gov.lv/de,embassy.germany@mfa.gov.lv
6,123811011,11001001,Botschaft Tadschikistan,52.529417,13.345070,POINT(13.345070470695365 52.5294172),Moabit,Mitte,0102,TJ,embassy,"Perleberger Straße 43, 10559 Berlin",http://www.botschaft-tadschikistan.de,info@botschaft-tadschikistan.de
7,25999440,11001001,Botschaft der Arabischen Republik Syrien,52.508277,13.350166,POINT(13.350165950673652 52.50827685),Tiergarten,Mitte,0104,SY,embassy,"Rauchstraße 25, 10787 Berlin",http://www.mofaex.gov.sy/berlin-embassy,<NA>
8,24045732,11001001,Botschaft der Arabischen Republik Ägypten,52.508669,13.363283,POINT(13.363283288286704 52.50866885),Tiergarten,Mitte,0104,EG,embassy,"Stauffenbergstraße 6-7, 10785 Berlin",https://egyptian-embassy.de/,embassy@egyptian-embassy.de
9,8615201458,11001001,Botschaft der Bolivarischen Republik Venezuela,52.503850,13.350180,POINT(13.3501797 52.5038501),Tiergarten,Mitte,0104,VE,embassy,"Schillstraße 10, 10785 Berlin",https://www.botschaft-venezuela.de,<NA>


In [39]:
df_missions_final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 214 entries, 0 to 213
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   id                214 non-null    int64  
 1   district_id       214 non-null    object 
 2   name              214 non-null    object 
 3   latitude          214 non-null    float64
 4   longitude         214 non-null    float64
 5   geometry          214 non-null    object 
 6   neighborhood      214 non-null    object 
 7   district          214 non-null    object 
 8   neighborhood_id   214 non-null    object 
 9   country_iso_code  214 non-null    object 
 10  mission_type      214 non-null    object 
 11  address           214 non-null    object 
 12  website           162 non-null    object 
 13  email             69 non-null     string 
dtypes: float64(2), int64(1), object(10), string(1)
memory usage: 23.5+ KB


## Validation

general data validation because this row looks strange:
1880737966	11004004	Malteser-Ritterorden	52.518571	13.311785	POINT(13.3117855 52.5185711)	Charlottenburg	Charlottenburg-Wilmersdorf	0401	embassy	SMOM

## CHECK 1 – Multiple or wrong embassies per country

All multiple sending countries were retained because they represent distinct diplomatic-related locations (e.g., main embassies, cultural centers, or specialized divisions) mapped separately in OSM, and therefore reflect real institutional presence rather than data duplication.

- Malaysia
    - -> MY
- SMOM (Sovereign Military Order of Malta)
    - SMOM represents the Sovereign Military Order of Malta, a recognized diplomatic entity under international law, and is therefore a legitimate non-state actor.
- LAS (League of Arab States)
    - LAS refers to the League of Arab States, an international organization with diplomatic representation, and is thus correctly included.
- AW;CW;NL;SX (Kingdom of the Netherlands – Caribbean parts)
    - The composite code AW;CW;NL;SX reflects the Kingdom of the Netherlands and its Caribbean constituent territories, representing a valid composite diplomatic designation rather than an error.

In [40]:
df_missions_final[
    ~df_missions_final["country_iso_code"].str.fullmatch(r"[A-Z]{2}", na=False)
]["country_iso_code"].unique()

array(['AW;CW;NL;SX', 'LAS', 'Malaysia', 'SMOM', 'Kenia', 'Luxembourg',
       'Haiti', 'Kongo'], dtype=object)

In [41]:
df_missions_final.loc[
    df_missions_final["country_iso_code"] == "Malaysia",
    "country_iso_code"
] = "MY"

In [42]:
df_missions_final[
    df_missions_final["country_iso_code"] == "AW;CW;NL;SX"
]

,id,district_id,name,latitude,longitude,geometry,neighborhood,district,neighborhood_id,country_iso_code,mission_type,address,website,email
112,515190251,11001001,Botschaft des Königreichs der Niederlande,52.515296,13.411963,POINT(13.4119633 52.5152959),Mitte,Mitte,0101,AW;CW;NL;SX,embassy,"Klosterstraße 50, 10179 Berlin",https://www.sieunddieniederlande.nl/,bln@minbuza.nl


In [43]:
# Count how many missions exist per country and mission type
# Normally, there should be only one embassy per country

(
    df_missions_final
    .groupby(["country_iso_code", "mission_type"])
    .size()
    .sort_values(ascending=False)
    .head(40)
)

country_iso_code  mission_type
CN                embassy         3
SZ                consulate       3
CL                embassy         2
SN                embassy         2
CU                embassy         2
CY                embassy         2
CZ                embassy         2
DJ                embassy         2
DK                embassy         2
DZ                embassy         2
ER                embassy         2
MC                embassy         2
GB                embassy         2
QA                embassy         2
GR                embassy         2
AE                embassy         2
IE                embassy         2
IL                embassy         2
IQ                embassy         2
IR                embassy         2
IS                embassy         2
NO                embassy         2
NE                embassy         2
KR                embassy         2
KW                embassy         2
MY                embassy         2
LB                embassy        

In [44]:
df_missions_final[
    (df_missions_final["country_iso_code"] == "CU") &
    (df_missions_final["mission_type"] == "embassy")
]

,id,district_id,name,latitude,longitude,geometry,neighborhood,district,neighborhood_id,country_iso_code,mission_type,address,website,email
52,42458597,11003003,Botschaft der Republik Kuba,52.556210,13.410943,POINT(13.41094313519865 52.5562097),Prenzlauer Berg,Pankow,0301,CU,embassy,"Stavangerstraße 20, 10439 Berlin",http://misiones.minrex.gob.cu/alemania,recepcion@botschaft-kuba.de
180,65104470,11003003,Residenz des Botschafters def Republik Kuba,52.555606,13.410849,POINT(13.410848722905271 52.55560625),Prenzlauer Berg,Pankow,0301,CU,embassy,"Ibsenstraße 12, 10439 Berlin",<NA>,<NA>


In [45]:
df_missions_final[
    (df_missions_final["country_iso_code"] == "IR") &
    (df_missions_final["mission_type"] == "embassy")
]

,id,district_id,name,latitude,longitude,geometry,neighborhood,district,neighborhood_id,country_iso_code,mission_type,address,website,email
17,33840382,11006006,Botschaft der Islamischen Republik Iran,52.460951,13.299962,POINT(13.299962054001433 52.46095135),Dahlem,Steglitz-Zehlendorf,0605,IR,embassy,"Podbielskiallee 67, 14195 Berlin",http://de.berlin.mfa.ir/,<NA>
158,12158159241,11006006,Iran Haus - Kulturabteilung der Botschaft der ...,52.432508,13.308956,POINT(13.308956 52.4325077),Lichterfelde,Steglitz-Zehlendorf,0602,IR,embassy,"Drakestraße, 12205 Berlin",<NA>,<NA>


In [46]:
df_missions_final[
    (df_missions_final["country_iso_code"] == "CN") &
    (df_missions_final["mission_type"] == "embassy")
]

,id,district_id,name,latitude,longitude,geometry,neighborhood,district,neighborhood_id,country_iso_code,mission_type,address,website,email
101,361488018,11001001,Botschaft der Volksrepublik China,52.513610,13.416802,POINT(13.4168021 52.5136105),Mitte,Mitte,0101,CN,embassy,"Märkisches Ufer 54, 10179 Berlin",http://www.china-botschaft.de/det/,<NA>
148,33755650,11001001,Chinesisches Kulturzentrum,52.507666,13.352558,POINT(13.352557849040139 52.507666150000006),Tiergarten,Mitte,0104,CN,embassy,"Klingelhöferstraße 21, 10785 Berlin",http://c-k-b.eu/,info@c-k-b.eu
205,5853422966,11003003,Wirtschafts- und Handelsabteilung der VR China...,52.575671,13.404782,POINT(13.4047822 52.5756715),Niederschönhausen,Pankow,0311,CN,embassy,"Majakowskiring 66, 13156 Berlin",<NA>,<NA>


## CHECK 2 – Suspicious or non-standard names

all ok

In [47]:
# Show entries whose names do NOT contain typical diplomatic keywords
# This helps detect cultural institutes, residences, historical objects, etc.

df_missions_final[
    ~df_missions_final["name"].str.contains(
        "Botschaft|Konsulat|Embassy|Consulate|Honor",
        case=False,
        na=False
    )
][["id", "name", "mission_type", "country_iso_code"]]

,id,name,mission_type,country_iso_code
1,39348994,Apostolische Nuntiatur in Deutschland,embassy,VA
148,33755650,Chinesisches Kulturzentrum,embassy,CN
167,479109936,Kulturzentrum Aserbaidschan,embassy,AZ
171,6939454627,Liga der Arabischen Staaten - Mission Berlin,consulate,LAS
174,1880737966,Malteser-Ritterorden,embassy,SMOM
175,78061449,Palästinensische Mission,embassy,PS
176,76778841,Republik Côte d’Ivoire,embassy,CI
177,77672466,Residence of the British ambassador,embassy,GB
178,10621330540,Residenz Dänemark,embassy,DK
200,78058642,Sendirad Islands,embassy,IS


## CHECK 3 – Duplicate coordinates

none -> ok


In [48]:
# Identify entries that share identical latitude and longitude
# These are often duplicate OSM node/way representations

df_missions_final[
    df_missions_final.duplicated(subset=["latitude", "longitude"], keep=False)
].sort_values(["latitude", "longitude"])

,id,district_id,name,latitude,longitude,geometry,neighborhood,district,neighborhood_id,country_iso_code,mission_type,address,website,email


## CHECK 4 – Mission type distribution

all ok

In [49]:
# Check which mission types exist in the dataset
df_missions_final["mission_type"].value_counts()

mission_type
embassy      200
consulate     14
Name: count, dtype: int64

In [50]:
df_missions_final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 214 entries, 0 to 213
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   id                214 non-null    int64  
 1   district_id       214 non-null    object 
 2   name              214 non-null    object 
 3   latitude          214 non-null    float64
 4   longitude         214 non-null    float64
 5   geometry          214 non-null    object 
 6   neighborhood      214 non-null    object 
 7   district          214 non-null    object 
 8   neighborhood_id   214 non-null    object 
 9   country_iso_code  214 non-null    object 
 10  mission_type      214 non-null    object 
 11  address           214 non-null    object 
 12  website           162 non-null    object 
 13  email             69 non-null     string 
dtypes: float64(2), int64(1), object(10), string(1)
memory usage: 23.5+ KB


## CHECK 5 - len country/ iso-code

In [51]:
# validate iso-code > 2 characters
bad = df_missions_final[
    df_missions_final["country_iso_code"].notna()
    & (df_missions_final["country_iso_code"].astype(str).str.len() != 2)
][["id", "name", "country_iso_code"]]

print(bad.head(30))
print("Anzahl bad:", len(bad))

              id                                          name  \
112    515190251     Botschaft des Königreichs der Niederlande   
171   6939454627  Liga der Arabischen Staaten - Mission Berlin   
174   1880737966                          Malteser-Ritterorden   
206  11091458818                           Botschaft von Kenia   
207  11687769898                      Botschaft von Luxembourg   
208  12541996229                           Botschaft von Haiti   
212     76978937                           Botschaft von Kongo   

    country_iso_code  
112      AW;CW;NL;SX  
171              LAS  
174             SMOM  
206            Kenia  
207       Luxembourg  
208            Haiti  
212            Kongo  
Anzahl bad: 7


In [52]:
# replace longer isocodes
replacements = {
    "Kenia": "KE",
    "Luxembourg": "LU",
    "Haiti": "HT",
    "Kongo": "CG",
    "AW;CW;NL;SX": "NL",
    "SMOM": None,
    "LAS": None
}

df_missions_final["country_iso_code"] = (
    df_missions_final["country_iso_code"]
    .replace(replacements)
)

In [53]:
# trim and uppercase for consistency
df_missions_final["country_iso_code"] = (
    df_missions_final["country_iso_code"]
    .apply(clean_str)
    .str.upper()
)

# double check validation
invalid = df_missions_final[
    df_missions_final["country_iso_code"].notna() &
    (df_missions_final["country_iso_code"].str.len() != 2)
]

print("Invalid remaining:", len(invalid))


Invalid remaining: 0


In [54]:
df_missions_final

,id,district_id,name,latitude,longitude,geometry,neighborhood,district,neighborhood_id,country_iso_code,mission_type,address,website,email
0,78061104,11004004,Afghanisches Konsulat,52.479652,13.278615,POINT(13.27861531770446 52.47965235),Grunewald,Charlottenburg-Wilmersdorf,0404,AF,consulate,"Kronberger Straße 5, 14193 Berlin",<NA>,<NA>
1,39348994,11008008,Apostolische Nuntiatur in Deutschland,52.488004,13.408922,POINT(13.408922272633824 52.488003649999996),Neukölln,Neukölln,0801,VA,embassy,"Lilienthalstraße 3a, 10965 Berlin",http://www.nuntiatur.de,<NA>
2,1880606692,11007007,Botschaft Argentinien,52.501211,13.344571,POINT(13.3445706 52.5012109),Schöneberg,Tempelhof-Schöneberg,0701,AR,embassy,"Kleiststraße 23-26, 10787 Berlin",http://www.ealem.mrecic.gov.ar/,<NA>
3,348075887,11004004,Botschaft Gabun,52.475747,13.314478,POINT(13.314478 52.4757469),Wilmersdorf,Charlottenburg-Wilmersdorf,0402,GA,embassy,"Hohensteiner Straße 16, 14197 Berlin",http://www.amba-allemagne.ga/,<NA>
4,40664508,11005005,Botschaft Jordanien,52.511867,13.201596,POINT(13.201596053308027 52.5118667),Wilhelmstadt,Spandau,0509,JO,embassy,"Heerstraße 201, 13595 Berlin",http://www.jordanembassy.de,<NA>
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
209,5580806145,11001001,Botschaft von DJ,52.504640,13.343969,POINT(13.3439691 52.5046401),Tiergarten,Mitte,0104,DJ,embassy,"Kurfürstenstraße 84, 10787 Berlin",http://www.dschibuti-botschaft.de/,<NA>
210,5580806144,11001001,Botschaft von LS,52.504692,13.343829,POINT(13.343829 52.5046918),Tiergarten,Mitte,0104,LS,embassy,"Kurfürstenstraße 84, 10787 Berlin",https://lesothoembassy.de/,info@lesothoembassy.de
211,64384567,11003003,Botschaft von ER,52.555918,13.411762,POINT(13.411761862347964 52.55591785),Prenzlauer Berg,Pankow,0301,ER,embassy,"Gotlandstraße 14, 10439 Berlin",<NA>,<NA>
212,76978937,11012012,Botschaft von Kongo,52.605031,13.221587,POINT(13.221587211175095 52.6050307),Heiligensee,Reinickendorf,1204,CG,embassy,13503 Berlin,<NA>,<NA>


# Database Upload

## Database Connection

In [ ]:
user_name=''
password=''

In [ ]:
# DB Connection
host = 'localhost'
port = '5433'
database = 'layereddb'
schema='berlin_source_data'

# connection to db after opening tunnel
engine = create_engine(f'postgresql+psycopg2://{user_name}:{password}@{host}:{port}/{database}')

In [57]:
print(engine.url)

postgresql+psycopg2://laura_peters:***@localhost:5433/layereddb


## Table Schema Definition

In [58]:
# Table Schema Definition
create_table_query = f"""
CREATE TABLE IF NOT EXISTS berlin_source_data.diplomatic_missions (
    id VARCHAR(20) PRIMARY KEY,
    district_id VARCHAR(20) NOT NULL,
    name VARCHAR(200) NOT NULL DEFAULT 'Unknown',
    latitude DECIMAL(9,6),
    longitude DECIMAL(9,6),
    geometry VARCHAR,
    neighborhood VARCHAR(100),
    district VARCHAR(100),
    neighborhood_id VARCHAR(20),
    country_iso_code CHAR(2), -- e.g., 'US', 'FR', 'JP'
    mission_type VARCHAR(50), -- Embassy, Consulate, Honorary Consulate
    address VARCHAR(200),
    website VARCHAR(200),
    email VARCHAR(100),
    CONSTRAINT district_id_fk FOREIGN KEY (district_id)
        REFERENCES berlin_source_data.districts(district_id)
        ON DELETE RESTRICT
        ON UPDATE CASCADE
);
"""

# Execute the query to create empty table
with engine.connect() as conn:
    conn.execute(text(create_table_query))
    conn.commit()  # commit the transaction


## DataFrame to the database

In [59]:
from sqlalchemy import text

query = """
SELECT column_name
FROM information_schema.columns
WHERE table_schema = :schema
  AND table_name = 'diplomatic_missions'
ORDER BY ordinal_position;
"""

with engine.connect() as conn:
    db_cols = [r[0] for r in conn.execute(text(query), {"schema": schema}).fetchall()]

print(db_cols)
print(df_missions_final.columns.tolist())


['id', 'district_id', 'name', 'latitude', 'longitude', 'geometry', 'neighborhood', 'district', 'neighborhood_id', 'country_iso_code', 'mission_type', 'address', 'website', 'email']
['id', 'district_id', 'name', 'latitude', 'longitude', 'geometry', 'neighborhood', 'district', 'neighborhood_id', 'country_iso_code', 'mission_type', 'address', 'website', 'email']


In [60]:
#  Send the DataFrame to the database using .to_sql()
df_missions_final.to_sql(
    'diplomatic_missions',      
    engine,
    schema=schema,
    if_exists='append', 
    index=False
)

print("DataFrame sent to PostgreSQL using .to_sql() with psycopg2!")

IntegrityError: (psycopg2.errors.UniqueViolation) duplicate key value violates unique constraint "diplomatic_missions_pkey"
DETAIL:  Key (id)=(78061104) already exists.

[SQL: INSERT INTO berlin_source_data.diplomatic_missions (id, district_id, name, latitude, longitude, geometry, neighborhood, district, neighborhood_id, country_iso_code, mission_type, address, website, email) VALUES (%(id__0)s, %(district_id__0)s, %(name_ ... 58239 characters truncated ... (country_iso_code__213)s, %(mission_type__213)s, %(address__213)s, %(website__213)s, %(email__213)s)]
[parameters: {'neighborhood__0': 'Grunewald', 'longitude__0': 13.27861531770446, 'address__0': 'Kronberger Straße 5, 14193 Berlin', 'website__0': None, 'geometry__0': 'POINT(13.27861531770446 52.47965235)', 'name__0': 'Afghanisches Konsulat', 'district__0': 'Charlottenburg-Wilmersdorf', 'neighborhood_id__0': '0404', 'email__0': None, 'mission_type__0': 'consulate', 'id__0': 78061104, 'district_id__0': '11004004', 'latitude__0': 52.47965235, 'country_iso_code__0': 'AF', 'neighborhood__1': 'Neukölln', 'longitude__1': 13.408922272633824, 'address__1': 'Lilienthalstraße 3a, 10965 Berlin', 'website__1': 'http://www.nuntiatur.de', 'geometry__1': 'POINT(13.408922272633824 52.488003649999996)', 'name__1': 'Apostolische Nuntiatur in Deutschland', 'district__1': 'Neukölln', 'neighborhood_id__1': '0801', 'email__1': None, 'mission_type__1': 'embassy', 'id__1': 39348994, 'district_id__1': '11008008', 'latitude__1': 52.488003649999996, 'country_iso_code__1': 'VA', 'neighborhood__2': 'Schöneberg', 'longitude__2': 13.3445706, 'address__2': 'Kleiststraße 23-26, 10787 Berlin', 'website__2': 'http://www.ealem.mrecic.gov.ar/', 'geometry__2': 'POINT(13.3445706 52.5012109)', 'name__2': 'Botschaft Argentinien', 'district__2': 'Tempelhof-Schöneberg', 'neighborhood_id__2': '0701', 'email__2': None, 'mission_type__2': 'embassy', 'id__2': 1880606692, 'district_id__2': '11007007', 'latitude__2': 52.5012109, 'country_iso_code__2': 'AR', 'neighborhood__3': 'Wilmersdorf', 'longitude__3': 13.314478, 'address__3': 'Hohensteiner Straße 16, 14197 Berlin', 'website__3': 'http://www.amba-allemagne.ga/', 'geometry__3': 'POINT(13.314478 52.4757469)', 'name__3': 'Botschaft Gabun', 'district__3': 'Charlottenburg-Wilmersdorf', 'neighborhood_id__3': '0402' ... 2896 parameters truncated ... 'district__210': 'Mitte', 'neighborhood_id__210': '0104', 'email__210': 'info@lesothoembassy.de', 'mission_type__210': 'embassy', 'id__210': 5580806144, 'district_id__210': '11001001', 'latitude__210': 52.5046918, 'country_iso_code__210': 'LS', 'neighborhood__211': 'Prenzlauer Berg', 'longitude__211': 13.411761862347964, 'address__211': 'Gotlandstraße 14, 10439 Berlin', 'website__211': None, 'geometry__211': 'POINT(13.411761862347964 52.55591785)', 'name__211': 'Botschaft von ER', 'district__211': 'Pankow', 'neighborhood_id__211': '0301', 'email__211': None, 'mission_type__211': 'embassy', 'id__211': 64384567, 'district_id__211': '11003003', 'latitude__211': 52.55591785, 'country_iso_code__211': 'ER', 'neighborhood__212': 'Heiligensee', 'longitude__212': 13.221587211175095, 'address__212': '13503 Berlin', 'website__212': None, 'geometry__212': 'POINT(13.221587211175095 52.6050307)', 'name__212': 'Botschaft von Kongo', 'district__212': 'Reinickendorf', 'neighborhood_id__212': '1204', 'email__212': None, 'mission_type__212': 'embassy', 'id__212': 76978937, 'district_id__212': '11012012', 'latitude__212': 52.6050307, 'country_iso_code__212': 'CG', 'neighborhood__213': 'Zehlendorf', 'longitude__213': 13.23070560862933, 'address__213': 'Niklasstraße 19, 14163 Berlin', 'website__213': 'https://www.srilanka-botschaft.de', 'geometry__213': 'POINT(13.230720560437014 52.4337453)', 'name__213': 'Botschaft von Sri Lanka', 'district__213': 'Steglitz-Zehlendorf', 'neighborhood_id__213': '0604', 'email__213': None, 'mission_type__213': 'embassy', 'id__213': 1162427498, 'district_id__213': '11006006', 'latitude__213': 52.433760425, 'country_iso_code__213': 'LK'}]
(Background on this error at: https://sqlalche.me/e/20/gkpj)

## Test Query

In [67]:
# query test data
query = f"SELECT * FROM {schema}.diplomatic_missions"

# Execute the query
with engine.connect() as conn:
    df= pd.read_sql(text(query), conn)
    conn.commit()  # commit the transaction
df

,id,district_id,name,latitude,longitude,geometry,neighborhood,district,neighborhood_id,country_iso_code,mission_type,address,website,email
0,78061104,11004004,Afghanisches Konsulat,52.479652,13.278615,POINT(13.27861531770446 52.47965235),Grunewald,Charlottenburg-Wilmersdorf,0404,AF,consulate,"Kronberger Straße 5, 14193 Berlin",None,None
1,39348994,11008008,Apostolische Nuntiatur in Deutschland,52.488004,13.408922,POINT(13.408922272633824 52.488003649999996),Neukölln,Neukölln,0801,VA,embassy,"Lilienthalstraße 3a, 10965 Berlin",http://www.nuntiatur.de,None
2,1880606692,11007007,Botschaft Argentinien,52.501211,13.344571,POINT(13.3445706 52.5012109),Schöneberg,Tempelhof-Schöneberg,0701,AR,embassy,"Kleiststraße 23-26, 10787 Berlin",http://www.ealem.mrecic.gov.ar/,None
3,348075887,11004004,Botschaft Gabun,52.475747,13.314478,POINT(13.314478 52.4757469),Wilmersdorf,Charlottenburg-Wilmersdorf,0402,GA,embassy,"Hohensteiner Straße 16, 14197 Berlin",http://www.amba-allemagne.ga/,None
4,40664508,11005005,Botschaft Jordanien,52.511867,13.201596,POINT(13.201596053308027 52.5118667),Wilhelmstadt,Spandau,0509,JO,embassy,"Heerstraße 201, 13595 Berlin",http://www.jordanembassy.de,None
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
209,5580806145,11001001,Botschaft von DJ,52.504640,13.343969,POINT(13.3439691 52.5046401),Tiergarten,Mitte,0104,DJ,embassy,"Kurfürstenstraße 84, 10787 Berlin",http://www.dschibuti-botschaft.de/,None
210,5580806144,11001001,Botschaft von LS,52.504692,13.343829,POINT(13.343829 52.5046918),Tiergarten,Mitte,0104,LS,embassy,"Kurfürstenstraße 84, 10787 Berlin",https://lesothoembassy.de/,info@lesothoembassy.de
211,64384567,11003003,Botschaft von ER,52.555918,13.411762,POINT(13.411761862347964 52.55591785),Prenzlauer Berg,Pankow,0301,ER,embassy,"Gotlandstraße 14, 10439 Berlin",None,None
212,76978937,11012012,Botschaft von Kongo,52.605031,13.221587,POINT(13.221587211175095 52.6050307),Heiligensee,Reinickendorf,1204,CG,embassy,13503 Berlin,None,None
